yugabyte/yugabyte-operator
# Add ability to safely scale down TServers

**Reference to design outline**: https://github.com/yugabyte/yugabyte-operator/issues/14#issuecomment-664153339

- Upgrade to Operator SDK 0.17.x
- Label pods from StatefulSet with `yugabyte.com/ybcluster-name` and
  cluster name as value.
- Watch events form pods having above label.
  - Triggers reconcile when a pod is annotated with
    `yugabyte.com/blacklist: true`.
- Watch events from StatefulSets created by the operator.
  - Any modifications to the StatefulSets will trigger reconcile.

- Add annotation `yugabyte.com/blacklist: true` to the pods which are
  going to be scaled down.
  - A synchronization loop adds or removes these pods to YB-Master's
  blacklist.

- Adds pkg/ybconfig for parsing JSON-encoded universe configuration.

- Add new status conditions.
  - `MovingData`: the value is `True` when there is a data move operation
    in progress. A synchronization loop takes care of updating this
    value.
  - `ScalingDownTServers`: `True` when scale down is triggered, becomes
    `False` once the StatefulSet is scaled down successfully.
- Introduce Status.TargetedTServerReplicas.
  - This keeps the value stored so that any other changes to spec gets
    ingored while data move or scale down is in progress.

- If TServers are scaling down then controller waits for the operation
  to get completed with an exponentially increasing retry interval.

- Uses exec.go from github.com/kanisterio/kanister/pkg/kube which is
  licensed under Apache License, Version 2.0.

Edit:
## How to test
1. Install the CRD and operator with,
   ```console
   $ kubectl apply -f https://raw.githubusercontent.com/bhavin192/yugabyte-operator/tserver-scale-down/deploy/crds/yugabyte.com_ybclusters_crd.yaml
   $ kubectl apply -f https://raw.githubusercontent.com/bhavin192/yugabyte-operator/tserver-example-yamls/deploy/operator.yaml
   ```
2. Create example-ybcluster
   ```console
   $ kubectl apply -f https://raw.githubusercontent.com/bhavin192/yugabyte-operator/tserver-example-yamls/deploy/crds/yugabyte.com_v1alpha1_ybcluster_cr.yaml
   ```
3. Increase the number of TServer (this step is equivalent to doing `kubectl edit ybcluster example-ybcluster`)
   ```console
   $ kubectl patch ybcluster example-ybcluster -nyb-operator --type='json' -p='[{"op": "replace", "path": "/spec/tserver/replicas", "value":5}]'
   ```
4. Do some operations on the database
   ```console
   $ kubectl run --image "yugabytedb/yb-sample-apps" \
     -n yb-operator --restart "Never" \
     java-client-cql1 -- --workload CassandraBatchKeyValue --nodes yb-tservers.yb-operator.svc.cluster.local:9042 --num_threads_write 4 --num_threads_read 6 --value_size 4096

   $ sleep 5m

   $ kubectl delete pod -n yb-operator java-client-cql1
   ```
5. Scale down the TServer replicas to 3 (equivalent to editing the example-ybcluster)
   ```console
   $ kubectl patch ybcluster example-ybcluster -nyb-operator --type='json' -p='[{"op": "replace", "path": "/spec/tserver/replicas", "value":3}]'
   ```
6. Check the number of replicas of yb-tservers StatefulSet (should be 5 as it will move the data out first)
   ```console
   $ kubectl get pods -n yb-operator
   ```
7. Check logs of the operator to see the progress of scale down operation
   ```console
   $ kubectl logs -n yb-operator -lname=yugabyte-operator --tail 200 --follow
   ```
